Signal/src/util/OWSBackupIO.m

```diff
@@ -0,0 +1,253 @@
+        return item;
+    }
+}
+
+#pragma mark - Decrypt
+
+- (BOOL)decryptFileAsFile:(NSString *)srcFilePath
+              dstFilePath:(NSString *)dstFilePath
+            encryptionKey:(NSData *)encryptionKey
+{
+    OWSAssert(srcFilePath.length > 0);
+    OWSAssert(encryptionKey.length > 0);
+
+    @autoreleasepool {
+
+        // TODO: Decrypt the file without loading it into memory.
+        NSData *data = [self decryptFileAsData:srcFilePath encryptionKey:encryptionKey];
+
+        if (data.length < 1) {
+            return NO;
+        }
+
+        NSError *error;
+        BOOL success = [data writeToFile:dstFilePath options:NSDataWritingAtomic error:&error];
+        if (!success || error) {
+            OWSProdLogAndFail(@"%@ error writing decrypted data: %@", self.logTag, error);
+            return NO;
+        }
+        [OWSFileSystem protectFileOrFolderAtPath:dstFilePath];
+
+        return YES;
+    }
+}
+
+- (nullable NSData *)decryptFileAsData:(NSString *)srcFilePath encryptionKey:(NSData *)encryptionKey
+{
+    OWSAssert(srcFilePath.length > 0);
+    OWSAssert(encryptionKey.length > 0);
+
+    @autoreleasepool {
+
+        if (![NSFileManager.defaultManager fileExistsAtPath:srcFilePath]) {
+            DDLogError(@"%@ missing downloaded file.", self.logTag);
+            return nil;
+        }
+
+        NSData *_Nullable srcData = [NSData dataWithContentsOfFile:srcFilePath];
+        if (srcData.length < 1) {
+            OWSProdLogAndFail(@"%@ could not load file into memory for decryption.", self.logTag);
+            return nil;
+        }
+
+        NSData *_Nullable dstData = [self decryptDataAsData:srcData encryptionKey:encryptionKey];
+        return dstData;
+    }
+}
+
+- (nullable NSData *)decryptDataAsData:(NSData *)encryptedData encryptionKey:(NSData *)encryptionKey
+{
+    OWSAssert(encryptedData);
+    OWSAssert(encryptionKey.length > 0);
+
+    @autoreleasepool {
+
+        // TODO: Decrypt the data using key;
+        NSData *unencryptedData = encryptedData;
+
+        return unencryptedData;
+    }
+}
+
+#pragma mark - Compression
+
+- (nullable NSData *)compressData:(NSData *)srcData
+{
+    OWSAssert(srcData);
+
+    @autoreleasepool {
+
+        if (!srcData) {
+            OWSProdLogAndFail(@"%@ missing unencrypted data.", self.logTag);
+            return nil;
+        }
+
+        size_t srcLength = [srcData length];
+        const uint8_t *srcBuffer = (const uint8_t *)[srcData bytes];
+        if (!srcBuffer) {
+            return nil;
+        }
+        // This assumes that dst will always be smaller than src.
+        //
+        // We slightly pad the buffer size to account for the worst case.
+        size_t dstBufferLength = srcLength + 64 * 1024;
+        uint8_t *dstBuffer = malloc(sizeof(uint8_t) * dstBufferLength);
+        if (!dstBuffer) {
+            return nil;
+        }
+        // TODO: Should we use COMPRESSION_LZFSE?
+        size_t dstLength
```

Comment:
I'm tempted to use COMPRESSION_LZFSE, an Apple-specific compression scheme offered by compressionlib that is supposedly better than the traditional alternatives.

Reply:
Speaking as a reformed bzip2/xzip user, I would guess there are some pretty diminishing returns on esoteric compression schemes, and that there may be a future benefit to using a more universal compression approach.

So, I'd vote for using something standard unless the wins are big, but I don't feel very strongly about it.

Reply:
10k messages + 5k attachments.

![simulator screen shot - iphone x - 2018-03-20 at 11 11 26](https://user-images.githubusercontent.com/625803/37663394-827704f8-2c2f-11e8-9347-f8a76641a0b1.png)

### With LZFSE

450kb database snapshot.

### With ZLIB

775kb database snapshot.

### With LZ4

998kb database snapshot.

### With LZMA

338kb database snapshot.

### Conclusion

LZMA has best performance AND is more widely adopted standard than LZFSE.

Reply:
Forgot to include one (irrelevant) test result.

### LZ4_RAW

905kb database snapshot.